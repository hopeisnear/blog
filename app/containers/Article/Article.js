import React, { PureComponent } from 'react';
import { func, shape } from 'prop-types';
import { Helmet } from 'react-helmet';
import { imageUrlFor } from 'utils/imageLoader';
import ArticleContent from 'components/ArticleContent';
import ShareArticle from 'components/ShareArticle';
import Comments from 'components/Comments/index';
import AddComment from 'containers/AddComment/Loadable';
import './Article.scss';

export default class Article extends PureComponent {
  componentDidMount() {
    !this.props.article && this.props.fetchArticle();
  }

  render() {
    const { article } = this.props;
    if (!article) {
      return null;
    }

    return (
      <div className="Article">
        <Helmet>
          <title>{article.title}</title>
          <meta name="description" content={article.teaser} />
          <meta property="og:site_name" content="GOOD dev" />
          <meta property="og:title" content={article.title} />
          <meta property="og:description" content={article.teaser} />
          <meta
            property="og:image"
            content={imageUrlFor(article.image)
              .ignoreImageParams()
              .width(360)
              .height(252)}
          />
          <meta property="og:image:type" content="image/png" />
          <meta property="og:image:width" content="360" />
          <meta property="og:image:height" content="252" />
        </Helmet>
        <div className="article__content">{<ArticleContent article={article} />}</div>
        <ShareArticle />
        {article.comments && <Comments comments={article.comments} commentsCount={article.commentsCount} />}
        <h3 className="article__reply_header">Leave a Reply</h3>
        <AddComment />
      </div>
    );
  }
}

Article.propTypes = {
  article: shape({}),
  fetchArticle: func.isRequired
};

Article.defaultProps = {
  article: undefined
};
