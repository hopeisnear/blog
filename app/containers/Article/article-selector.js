import { createSelector } from 'reselect';

const selectArticle = () => createSelector(
  (state) => state.get('article'),
  (article) => article.get('article')
);

export { selectArticle };
