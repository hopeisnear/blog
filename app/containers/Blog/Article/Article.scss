@import "app/styles/variables.scss";

.blog-article {
  display: flex;
  justify-content: space-between;
  flex-direction: column;
  padding-bottom: 60px;

  a {
    text-decoration: underline;
    color: $active-link-color;
  }

  .blog-article-content {
    margin-bottom: 60px;
  }

  .comment-form-header {
    border: 0;
    margin-bottom: 0;
  }
}
