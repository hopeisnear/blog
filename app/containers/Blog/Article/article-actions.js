import { fetchArticle } from 'services/article-service';

export const FETCH_ARTICLE_RESPONDED = 'FETCH_ARTICLE_RESPONDED';

export function fetchArticleAction(articleName) {
  return (dispatch) => fetchArticle(articleName)
    .then((article) => {
      dispatch({ type: FETCH_ARTICLE_RESPONDED, article });
    });
}
