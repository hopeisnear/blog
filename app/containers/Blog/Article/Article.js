import React, { PureComponent } from 'react';
import { func, string, shape } from 'prop-types';
import { Helmet } from 'react-helmet';
import ArticleContent from 'components/ArticleContent';
import Comments from 'components/Comments';
import AddComment from 'containers/Blog/AddComment';

import './Article.scss';

export default class Article extends PureComponent {
  componentDidMount() {
    if (!this.props.article) {
      this.props.fetchArticle(this.props.articleName);
    }
  }

  render() {
    return (
      <div className="Article">
        <Helmet>
          <title>Article details</title>
          <meta
            name="description"
            content={`Article ${this.props.articleName} content`}
          />
        </Helmet>
        <div className="article__content">
          {this.props.article && <ArticleContent article={this.props.article} />}
        </div>
        {this.props.article && this.props.article.comments && <Comments comments={this.props.article.comments} />}
        <h3 className="article__reply_header">Leave a Reply</h3>
        <AddComment />
      </div>
    );
  }
}

Article.propTypes = {
  article: shape({}),
  articleName: string.isRequired,
  fetchArticle: func.isRequired
};

Article.defaultProps = {
  article: undefined
};
