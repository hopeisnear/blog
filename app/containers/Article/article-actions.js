import { fetchArticle } from 'services/article-service';
import { FETCH_ARTICLE_RESPONDED } from 'common/article-actions';
import { selectArticleName } from './article-selector';

export function fetchArticleAction() {
  return (dispatch, getState) =>
    fetchArticle(selectArticleName()(getState())).then(article => {
      dispatch({ type: FETCH_ARTICLE_RESPONDED, article });
    });
}
