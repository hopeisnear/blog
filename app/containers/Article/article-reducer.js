import { fromJS } from 'immutable';
import { FETCH_ARTICLE_REQUESTED, FETCH_ARTICLE_RESPONDED } from './article-actions';
import { createReducer } from '../../utils/create-reducer';

const initialState = fromJS({
  article: {}
});

export const articleReducer = createReducer(initialState, {
  [FETCH_ARTICLE_REQUESTED](state) {
    return state.set('article', fromJS({}));
  },
  [FETCH_ARTICLE_RESPONDED](state, action) {
    return state.set('article', fromJS(action.response));
  }
});

export default articleReducer;
