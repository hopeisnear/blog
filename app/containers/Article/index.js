import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';
import injectReducer from 'utils/injectReducer';

import { fetchArticleAction } from './article-actions';
import { selectArticle } from './article-selector';
import articleReducer from './article-reducer';
import Article from './Article';

const mapDispatchToProps = (dispatch) => ({
  fetchArticle: () => dispatch(fetchArticleAction({ id: 'test' }))
});

const mapStateToProps = createStructuredSelector({
  article: selectArticle()
});

const withConnect = connect(mapStateToProps, mapDispatchToProps);

const withReducer = injectReducer({ key: 'article', reducer: articleReducer });

export default compose(withReducer, withConnect)(Article);
export { mapDispatchToProps };
