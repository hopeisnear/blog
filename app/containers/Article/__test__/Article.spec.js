import React from 'react';
import { shallow } from 'enzyme';
import { spy, assert } from 'sinon';

import Article from '../Article';

describe('<Article />', () => {
  test('should render <Article />', () => {
    const articleComponent = shallow(<Article {...createProps()} />);

    expect(articleComponent).toMatchSnapshot();
  });

  test('should fetch new article on init', () => {
    const articleComponent = shallow(<Article {...createProps()} article={undefined} />);

    assert.calledOnce(articleComponent.instance().props.fetchArticle);
  });

  function createProps() {
    return {
      articleName: 'Test article',
      article: {
        comments: [],
        commentsCount: 2,
        title: 'article title',
        teaser: 'article teaser',
        image: ''
      },
      fetchArticle: spy()
    };
  }
});
