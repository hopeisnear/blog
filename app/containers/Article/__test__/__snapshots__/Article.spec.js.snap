// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Article /> should render <Article /> 1`] = `
<div
  className="Article"
>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title />
    <meta
      name="description"
    />
    <meta
      content="GOOD dev"
      property="og:site_name"
    />
    <meta
      property="og:title"
    />
    <meta
      property="og:description"
    />
    <meta
      content={
        ImageUrlBuilder {
          "options": Object {
            "baseUrl": "https://cdn.sanity.io",
            "dataset": "production",
            "height": 252,
            "ignoreImageParams": true,
            "projectId": "klyyehyz",
            "source": undefined,
            "width": 360,
          },
        }
      }
      property="og:image"
    />
    <meta
      content="image/png"
      property="og:image:type"
    />
    <meta
      content="360"
      property="og:image:width"
    />
    <meta
      content="252"
      property="og:image:height"
    />
  </HelmetWrapper>
  <div
    className="article__content"
  >
    <ArticleContent
      article={
        Object {
          "comments": Array [],
          "commentsCount": 2,
        }
      }
    />
  </div>
  <ShareArticle />
  <Comments
    comments={Array []}
    commentsCount={2}
  />
  <h3
    className="article__reply_header"
  >
    Leave a Reply
  </h3>
  <LoadableComponent />
</div>
`;
